TypeCobol.Analysis/Dfa/DataFlowGraphBuilder.cs

```diff
@@ -0,0 +1,486 @@
+        /// </summary>
+        public ControlFlowGraph<N, DfaBasicBlockInfo<V>> Cfg
+        {
+            get;
+            internal set;
+        }
+
+        /// <summary>
+        /// Constructor
+        /// </summary>
+        /// <param name="cfg">Underlying Control Flow Graph</param>
+        protected DataFlowGraphBuilder(ControlFlowGraph<N, DfaBasicBlockInfo<V>> cfg)
+        {
+            System.Diagnostics.Debug.Assert(cfg != null);
+            this.Cfg = cfg;
+        }
+
+        /// <summary>
+        /// The USE List
+        /// </summary>
+        public List<DfaUsePoint<N, V>> UseList
+        {
+            get;
+            internal set;
+        }
+
+        /// <summary>
+        /// The DEF List
+        /// </summary>
+        public List<DfaDefPoint<N, V>> DefList
+        {
+            get;
+            internal set;
+        }
+
+        /// <summary>
+        /// The dictionary which gives for each Variable defined its Def set.
+        /// That is to say all definitions of this variable.
+        /// This Dictionary id built during the GEN set computation.
+        /// </summary>
+        public Dictionary<V, BitSet> VariableDefMap
+        {
+            get;
+            internal set;
+        }
+
+        /// <summary>
+        /// The Definitions counter
+        /// </summary>
+        private int _defCounter = 0;
+
+        /// <summary>
+        /// Get Use variables for a given Node
+        /// </summary>
+        /// <param name="node"></param>
+        /// <returns>The USE variable set</returns>
+        public abstract HashSet<V> GetUseVariables(N node);
+
+        /// <summary>
+        /// Get Def variables for a given Node
+        /// </summary>
+        /// <param name="node"></param>
+        /// <returns>The USE variable set</returns>
+        public abstract HashSet<V> GetDefVariables(N node);
+
+        /// <summary>
+        /// On Use Point Delegate
+        /// </summary>
+        /// <param name="dfaBuilder">The Dfa Builder in which the Use Point is seen</param>
+        /// <param name="usePoint">The Use point</param>
+        public delegate void OnUsePoint(DataFlowGraphBuilder<N, V> dfaBuilder, DfaUsePoint<N, V> usePoint);
+
+        /// <summary>
+        /// On Use Point Event
+        /// </summary>
+        public event OnUsePoint  OnUsePointEvent;
+
+        /// <summary>
+        /// On Def Point Delegate
+        /// </summary>
+        /// <param name="dfaBuilder">The Dfa Builder in which the Def Point is seen</param>
+        /// <param name="defPoint">The Def Point</param>
+        public delegate void OnDefPoint(DataFlowGraphBuilder<N, V> dfaBuilder, DfaDefPoint<N, V> defPoint);
+
+        /// <summary>
+        /// On Def Point Event
+        /// </summary>
+        public event OnDefPoint OnDefPointEvent;
+
+        /// <summary>
+        /// Compute the Use List
+        /// </summary>
+        public void ComputeUseList()
+        {
+            lock (this)
```

Comment:
The lock here is to prevent any other thread to build DFA when another is building it. If it is not intended to be build in an multi thread environment, ok, I agree.

Reply:
locks removed